csrc/multidevice/communication.cpp

```diff
@@ -107,214 +209,132 @@ std::string Communication::toString(int indent) const {
-    }
-  } else {
-    assertBufferCount(params_.src_bufs, 0);
-    assertBufferCount(params_.dst_bufs, 1);
-  }
-
-  if (params_.team.size() == 1) {
-    return nullptr;
-  }
-
-  return comm.getBackendForTeam(params_.team, backend)
-      ->broadcast(
-          comm.deviceId() == params_.root ? params_.src_bufs : params_.dst_bufs,
-          {.rootRank = root_relative_index_});
-}
-
-Gather::Gather(CommParams params) : Communication(params, "gather") {
-  assertBufferCount(params_.src_bufs, 1);
+std::string Communication::toInlineString(int indent_size) const {
+  return toString(indent_size);
 }
 
-c10::intrusive_ptr<c10d::Work> Gather::post(
-    Communicator& comm,
-    std::optional<CommunicatorBackend> backend) {
-  post_common(*this, comm);
-  // This is used to change the representation of the buffers to match c10d
-  // ProcessGroup API
-  std::vector<std::vector<at::Tensor>> buf_list;
-  if (comm.deviceId() == params_.root) {
-    assertBufferCount(params_.dst_bufs, params_.team.size());
-    buf_list = {std::move(params_.dst_bufs)};
-  } else {
-    assertBufferCount(params_.dst_bufs, 0);
+// TODO add checking symbolic representation of src and dst buffers
+bool Communication::sameAs(const Statement* other) const {
+  if (other == this) {
+    return true;
   }
-  auto work =
-      comm.getBackendForTeam(params_.team, backend)
-          ->gather(
-              buf_list, params_.src_bufs, {.rootRank = root_relative_index_});
-  if (comm.deviceId() == params_.root) {
-    params_.dst_bufs = std::move(buf_list.back());
+  if (!other->isA<Communication>()) {
+    return false;
   }
-  return work;
-}
-
-Allgather::Allgather(CommParams params)
-    : Communication(params, "allgather", false) {
-  assertBufferCount(params_.src_bufs, 1);
-  assertBufferCount(params_.dst_bufs, params_.team.size());
-}
-
-c10::intrusive_ptr<c10d::Work> Allgather::post(
-    Communicator& comm,
-    std::optional<CommunicatorBackend> backend) {
-  post_common(*this, comm);
-  // This is used to change the representation of the buffers to match c10d
-  // ProcessGroup API
-  std::vector<std::vector<at::Tensor>> buf_list;
-  buf_list = {std::move(params_.dst_bufs)};
-  auto work = comm.getBackendForTeam(params_.team, backend)
-                  ->allgather(buf_list, params_.src_bufs, {});
-  params_.dst_bufs = std::move(buf_list.back());
-  return work;
-}
+  const auto& p1 = this->params();
+  const auto& p2 = other->as<Communication>()->params();
 
-Scatter::Scatter(CommParams params) : Communication(params, "scatter") {
-  assertBufferCount(params_.dst_bufs, 1);
+  return (
+      p1.type == p2.type && (!hasRoot(p1.type) || p1.root == p2.root) &&
+      p1.team == p2.team && (!isReduction(p1.type) || p1.redOp == p2.redOp));
 }
 
-c10::intrusive_ptr<c10d::Work> Scatter::post(
-    Communicator& comm,
-    std::optional<CommunicatorBackend> backend) {
-  post_common(*this, comm);
+c10::intrusive_ptr<c10d::Work> postSingleCommunication(
```

Comment:
Just confirming, this is moving all the <Collective>::post logic into `postSingleCommunication` with the logic remaining the same?
Minus all validation checks which got moved into their own function.

Reply:
Yes, there is no functional change in this PR, only some refactor and reorganization. The checks have all been moved to an auxiliary function `assertValid`